proxy/pkg/zdmproxy/streamidmapper.go

```diff
@@ -0,0 +1,130 @@
+package zdmproxy
+
+import (
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"sync"
+)
+
+var maxStreamIds int16 = 2048
+
+// StreamIdMapper is used to map the incoming stream ids from the client/driver to internal ids managed by the proxy
+// This is required because we also generate requests inside the proxy that go to the cluster through the same connection,
+// hence they must have non-conflicting ids with user's requests.
+type StreamIdMapper interface {
+	GetNewIdFor(streamId int16) (int16, error)
+	RestoreId(syntheticId int16) (int16, error)
+	ReleaseId(syntheticId int16) (int16, error)
+}
+
+type cqlStreamIdMapper struct {
+	sync.Mutex
+	ids chan int16
+}
+
+func NewCqlStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	var ids = make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		ids <- i
+	}
+	return &cqlStreamIdMapper{
+		ids: ids,
+	}
+}
+
+func (icsim *cqlStreamIdMapper) GetNewIdFor(_ int16) (int16, error) {
+	icsim.Lock()
+	defer icsim.Unlock()
+	select {
+	case id, ok := <-icsim.ids:
+		if ok {
+			return id, nil
+		} else {
+			return -1, fmt.Errorf("stream id channel closed")
+		}
+	default:
+		return -1, fmt.Errorf("no stream id available")
+	}
+}
+
+func (icsim *cqlStreamIdMapper) RestoreId(_ int16) (int16, error) {
+	return -1, fmt.Errorf("internal CQL ids have no original ids to be restored")
```

Comment:
This should always return the original id even though this mapper is essentially not doing any mapping because from the outside it can be used as if it was mapping (which is kind of the contract of the interface)